File: crates/hyper_engine/src/engine.rs
//
// Copyright (c) 2022-2024, SkillerRaptor
//
// SPDX-License-Identifier: MIT
//

use std::{sync::Arc, time::Instant};

use hyper_math::Vec2;
use hyper_platform::{Event, EventBus, Input, Window, WindowDescriptor};
use hyper_render::{
    graphics_device::{GraphicsApi, GraphicsDevice, GraphicsDeviceDescriptor},
    renderer::Renderer,
    scene::Scene,
    surface::{Surface, SurfaceDescriptor},
};
use raw_window_handle::{HasDisplayHandle, HasWindowHandle};

#[derive(Clone, Debug)]
pub struct EngineDescriptor {
    pub width: u32,
    pub height: u32,
    pub graphics_api: GraphicsApi,
    pub debug: bool,
}

pub struct Engine {
    scene: Scene,

    // NOTE: Maybe adding multiple kinds of renderer?
    renderer: Renderer,
    surface: Box<dyn Surface>,
    graphics_device: Arc<dyn GraphicsDevice>,

    window: Window,

    input: Input,
    event_bus: EventBus,
}

impl Engine {
    pub fn new(descriptor: &EngineDescriptor) -> Self {
        let start_time = Instant::now();

        let event_bus = EventBus::default();
        let input = Input::default();

        // Assuming that the engine will run in editor-mode
        // FIXME: When adding game mode change the title to the game
        let title = if cfg!(debug_assertions) {
            "Hyper Engine (Debug Build)"
        } else {
            "Hyper Engine"
        };

        let window = Window::new(&WindowDescriptor {
            title: &title,
            width: descriptor.width,
            height: descriptor.height,
        })
        .unwrap();

        let graphics_device =
            hyper_render::graphics_device::create_graphics_device(&GraphicsDeviceDescriptor {
                graphics_api: descriptor.graphics_api,
                debug_mode: descriptor.debug,
                display_handle: window.display_handle().unwrap(),
            });

        let surface = graphics_device.create_surface(&SurfaceDescriptor {
            display_handle: window.display_handle().unwrap(),
            window_handle: window.window_handle().unwrap(),
            window_size: window.inner_size(),
        });

        let renderer = Renderer::new(&graphics_device);

        // NOTE: This is temporary till real scene switch will be implemented
        let scene = Scene::new();

        tracing::info!(
            "Engine initialized in {:.2} seconds",
            start_time.elapsed().as_secs_f32()
        );

        Self {
            scene,

            renderer,
            surface,
            graphics_device,

            window,

            input,
            event_bus,
        }
    }

    pub fn run(&mut self) {
        // Fixed at 60 frames per second
        let mut _time = 0.0;
        let delta_time = 1.0 / 60.0;

        let mut current_time = Instant::now();
        let mut accumulator = 0.0;

        'running: loop {
            let new_time = Instant::now();
            let frame_time = (new_time - current_time).as_secs_f32();
            current_time = new_time;

            accumulator += frame_time;

            self.window.process_events(&mut self.event_bus);

            while let Some(event) = self.event_bus.pop_event() {
                match event {
                    Event::MouseButtonPressed { mouse_code } => {
                        self.input.set_mouse_state(mouse_code, true)
                    }
                    Event::MouseButtonReleased { mouse_code } => {
                        self.input.set_mouse_state(mouse_code, false)
                    }
                    Event::MouseMoved { x, y } => self
                        .input
                        .set_cursor_position(Vec2::new(x as f32, y as f32)),
                    Event::KeyPressed { key_code } => self.input.set_key_state(key_code, true),
                    Event::KeyReleased { key_code } => self.input.set_key_state(key_code, false),
                    Event::WindowClose => break 'running,
                    Event::WindowResize { width, height } => {
                        let width = width.max(1);
                        let height = height.max(1);
                        self.surface.resize(width, height);
                    }
                    _ => {}
                }
            }

            while accumulator >= delta_time {
                // Update Fixed
                accumulator -= delta_time;
                _time += delta_time;
            }

            // Update

            // Render
            self.renderer.render(&mut self.surface, &self.scene);
        }

        self.graphics_device.wait_for_idle();
    }
}

File: hyper_engine/src/hyper_engine/engine.cpp
/*
 * Copyright (c) 2024, SkillerRaptor
 *
 * SPDX-License-Identifier: MIT
 */

#include "hyper_engine/engine.hpp"

#include <chrono>

#include <hyper_core/assertion.hpp>
#include <hyper_core/logger.hpp>
#include <hyper_core/prerequisites.hpp>

namespace hyper_engine
{
    Engine::Engine(const EngineDescriptor &descriptor)
        : m_start_time(std::chrono::steady_clock::now())
        , m_running(false)
        , m_window({
              .title = "HyperEngine",
              .width = descriptor.width,
              .height = descriptor.height,
              .event_bus = m_event_bus,
          })
        , m_graphics_device(hyper_rhi::GraphicsDevice::create({
              .graphics_api = descriptor.graphics_api,
              .debug_mode = descriptor.debug,
          }))
        , m_surface(m_graphics_device->create_surface({
              .window = m_window,
          }))
        , m_renderer({
              .graphics_device = m_graphics_device,
              .surface = m_surface,
          })
    {
        HE_ASSERT(m_graphics_device);

        m_event_bus.subscribe<hyper_platform::WindowCloseEvent>(HE_BIND_FUNCTION(Engine::on_close));
        m_event_bus.subscribe<hyper_platform::WindowResizeEvent>(HE_BIND_FUNCTION(Engine::on_resize));

        m_running = true;

        const std::chrono::steady_clock::time_point end_time = std::chrono::steady_clock::now();
        const std::chrono::duration<double> elapsed_seconds = end_time - m_start_time;

        HE_INFO("Engine initialized in {:.2}s", elapsed_seconds.count());
    }

    void Engine::run()
    {
        // float time = 0.0;
        constexpr auto delta_time = static_cast<float>(1.0 / 60.0);

        std::chrono::time_point current_time = std::chrono::steady_clock::now();
        float accumulator = 0.0;

        while (m_running)
        {
            const std::chrono::time_point new_time = std::chrono::steady_clock::now();
            const float frame_time = std::chrono::duration<float>(new_time - current_time).count();
            current_time = new_time;

            accumulator += frame_time;

            hyper_platform::Window::poll_events();

            while (accumulator >= delta_time)
            {
                // Fixed Update

                accumulator -= delta_time;
                // time += delta_time;
            }

            // Update

            // Render
        }
    }

    void Engine::on_close(const hyper_platform::WindowCloseEvent &)
    {
        m_running = false;
    }

    void Engine::on_resize(const hyper_platform::WindowResizeEvent &event)
    {
        HE_DEBUG("{}, {}", event.width(), event.height());
    }
} // namespace hyper_engine
